rskj-core/src/main/java/org/ethereum/datasource/DataSourceWithCache.java

```diff
@@ -0,0 +1,146 @@
+import java.util.stream.Stream;
+
+public class DataSourceWithCache implements KeyValueDataSource {
+    private final KeyValueDataSource base;
+    private final Map<ByteArrayWrapper, byte[]> uncommittedCacheMap;
+    private final Map<ByteArrayWrapper, byte[]> committedCacheMap;
+
+    // During the processing of a Raul's fully filled blockchain, the cache
+    // is generating the following hits per block (average)
+    // uncommitedCacheHits = 1340
+    // commitedCacheHits = 248
+    // Processing 134 blocks grows the commitedCache to 100K entries, or approximately
+    // or 25 Mbytes. A cache of about 100 Mbytes seems rasonable. Anyway, for
+    // precaution we'll set the limit to 100K entries.
+    public DataSourceWithCache(KeyValueDataSource base, int cacheSize) {
+        this.base = base;
+        this.uncommittedCacheMap = new MaxSizeHashMap<>(cacheSize, true);
+        this.committedCacheMap = new MaxSizeHashMap<>(cacheSize, true);
+    }
+
+    @Override
+    public byte[] get(byte[] key) {
+        ByteArrayWrapper wrappedKey = ByteUtil.wrap(key);
+        byte[] value;
+        value = committedCacheMap.get(wrappedKey);
+        if (value != null) {
+            return value;
+        }
+
+        value = uncommittedCacheMap.get(wrappedKey);
+        if (value != null) {
+            return value;
+        }
+
+        value = base.get(key);
+        committedCacheMap.put(wrappedKey, value);
+        return value;
+    }
+
+    @Override
+    public byte[] put(byte[] key, byte[] value) {
+        Objects.requireNonNull(key);
+        Objects.requireNonNull(value);
+        ByteArrayWrapper wrappedKey = ByteUtil.wrap(key);
+
+        // here I could check for equal datas or just move to the uncommited uncommittedCache.
+        byte[] priorValue = committedCacheMap.get(wrappedKey);
+        if (priorValue != null && Arrays.equals(priorValue, value)) {
+            return value;
+        }
+
+        committedCacheMap.remove(wrappedKey);
+        uncommittedCacheMap.put(wrappedKey, value);
+        return value;
+    }
+
+    @Override
+    public void delete(byte[] key) {
+        Objects.requireNonNull(key);
+        ByteArrayWrapper wrappedKey = ByteUtil.wrap(key);
+        // fully delete this element
+        committedCacheMap.remove(wrappedKey);
+        uncommittedCacheMap.remove(wrappedKey);
+    }
+
+    @Override
+    public Set<byte[]> keys() {
+        Stream<ByteArrayWrapper> baseKeys = base.keys().stream().map(ByteArrayWrapper::new);
+        Stream<ByteArrayWrapper> committedKeys = committedCacheMap.entrySet().stream()
+                .filter(e -> e.getValue() == null)
+                .map(Map.Entry::getKey);
+        // there is no need to filter null values on uncommittedCacheMap
+        Stream<ByteArrayWrapper> uncommittedKeys = uncommittedCacheMap.entrySet().stream()
+                .map(Map.Entry::getKey);
+
+        return Stream.concat(Stream.concat(baseKeys, committedKeys), uncommittedKeys)
```

Comment:
[BUG] You shouldn't map and then collect because `byte[]`s doesn't implement deep equality

Reply:
Great catch, thanks!